File: MD.CMS.BusinessLogic.Core/DataAccess/Entities/Content.cs
using MD.CMS.BusinessLogic.Core.Properties;
using System;
using System.Collections.Generic;
using System.Linq;
using MD.Tools.BaseDataAccess.Core.Entities;
using MD.Tools.BaseDataAccess.Plugins.Core.Mapping.ReportDefinition.ReportDesigner.CustomAttributes;
using System.Diagnostics.CodeAnalysis;

namespace MD.CMS.BusinessLogic.Core.DataAccess.Entities
{
    public class Content : BaseEntity<string>, IComparable<Content>
    {
        #region Attributes
        private int _LCID;
        private DateTime _dateCreated;
        private string _authorId;
        private long _folderId;
        private string _title;
        private string _path;
        private string _html;
        private User _author;
        private ContentTypeDefinition<ContentTypeDefinitionFieldValue> _contentType;
        private List<MetaDataFieldValue> _metaDataFieldValues;
        private List<Taxonomy> _taxonomy;
        private List<Menu> _menu;
        private long _taxonomyId;
        private List<ContentAlias> _contentAliases;
        private long _contentTypeDefinitionId;
        private Template _template;
        private bool _isPublished;
        private bool _approvalPending;
        private bool? _isNew;
        #endregion

        #region Properties

        public int LCID
        {
            get
            {
                if (_LCID.Equals(default(int)))
                {
                    _LCID = Settings.Default.DefaultLcid;
                }
                return _LCID;
            }
            set { _LCID = value; }
        }

        public string DateCreated
        {
            get
            {
                return _dateCreated.ToString("yyyy-MM-dd HH:mm:ss");
            }
            set
            {
                if (!string.IsNullOrEmpty(value))
                {
                    _dateCreated = DateTime.Parse(value);
                }
                else
                {
                    _dateCreated = DateTime.UtcNow;
                }

            }
        }

        public bool IsDataBound
        {
            get
            {
                return ContentType != null && ContentType.Fields.Any(f => f.DataBound);
            }
        }

        [OmitPropertyFromReport]
        public List<ContentAlias> ContentAliases
        {
            get
            {
                if (_contentAliases == null)
                {
                    _contentAliases = new List<ContentAlias>();
                }
                return _contentAliases;
            }
            set { _contentAliases = value; }
        }
        public string AuthorId
        {
            get { return _authorId; }
            set { _authorId = value; }
        }
        public long ContentTypeDefinitionId
        {
            get { return _contentTypeDefinitionId; }
            set { _contentTypeDefinitionId = value; }
        }
        public long FolderId
        {
            get { return _folderId; }
            set { _folderId = value; }
        }
        public long TaxonomyId
        {
            get { return _taxonomyId; }
            set { _taxonomyId = value; }
        }
        public string Title
        {
            get { return _title; }
            set { _title = value; }
        }
        public string Path
        {
            get { return _path; }
            set { _path = value; }
        }
        public string UniqueId
        {
            get
            {
                if (!this.IsNew)
                {
                    return string.Format("{0}-{1}", this.Id, this.LCID);
                }
                return string.Empty;
            }
        }

		//[JsonIgnore : JsonNotIgnore]      
		public string Html
		{
			get { return _html; }
			set { _html = value; }
		}
		[OmitPropertyFromReport]
		public User Author
		{
			get { return _author; }
			set { _author = value; }
		}

		[OmitPropertyFromReport]
		public ContentTypeDefinition<ContentTypeDefinitionFieldValue> ContentType
		{
			get { return _contentType; }
			set { _contentType = value; }
		}

		[OmitPropertyFromReport]
		public bool IsNew
		{
			get
			{
                if(_isNew != null)
                {
                    return _isNew.Value;
                }
				return string.IsNullOrEmpty(Id);
			}
            set
            {
                _isNew = value;
            }
		}

		[OmitPropertyFromReport]
		public List<Taxonomy> Taxonomy
		{
			get { return _taxonomy; }
			set { _taxonomy = value; }
		}

		[OmitPropertyFromReport]
		public List<Menu> Menu
		{
			get { return _menu; }
			set { _menu = value; }
		}

		[OmitPropertyFromReport]
		public List<MetaDataFieldValue> MetaDataFieldValues
		{
			get { return _metaDataFieldValues; }
			set { _metaDataFieldValues = value; }
		}

		[OmitPropertyFromReport]
		public Template Template
		{
			get { return _template; }
			set { _template = value; }
		}
		/// <summary>
		/// Is Published?
		/// </summary>
		public bool IsPublished
		{
			get { return _isPublished; }
			set { _isPublished = value; }
		}
		public bool ApprovalPending
		{
			get { return _approvalPending; }
			set { _approvalPending = value; }
		}
		#endregion

		#region Methods
        public Content() : base()
        {

        }

        public Content(Content obj) : base(obj)
        {
            DateCreated = obj.DateCreated;
            LCID = obj.LCID;
            ContentAliases = obj.ContentAliases;
            AuthorId = obj.AuthorId;
            Author = obj.Author;
            ContentTypeDefinitionId = obj.ContentTypeDefinitionId;
            ContentType = obj.ContentType;
            FolderId = obj.FolderId;
            TaxonomyId = obj.TaxonomyId;
            Title = obj.Title;
            Path = obj.Path;
            Html = obj.Html;
            IsNew = obj.IsNew;
            Taxonomy = obj.Taxonomy;
            Menu = obj.Menu;
            MetaDataFieldValues = obj.MetaDataFieldValues;
            Template = obj.Template;
            IsPublished = obj.IsPublished;
            ApprovalPending = obj.ApprovalPending;
        }

		public virtual bool ShouldSerializeLCID()
		{
			return true;
		}

		public virtual bool ShouldSerializeDateCreated()
		{
			return true;
		}

		public virtual bool ShouldSerializeAuthorId()
		{
			return true;
		}

		public virtual bool ShouldSerializeAuthor()
		{
			return true;
		}

		public virtual bool ShouldSerializeFolderId()
		{
			return true;
		}

		public virtual bool ShouldSerializeTaxonomyId()
		{
			return true;
		}

		public virtual bool ShouldSerializeTitle()
		{
			return true;
		}
		public virtual bool ShouldSerializePath()
		{
			return true;
		}

		public virtual bool ShouldSerializeHtml()
		{
			return true;
		}

		public virtual bool ShouldSerializeContentType()
		{
			return true;
		}

		public virtual bool ShouldSerializeContentAlias()
		{
			return true;
		}

		public virtual bool ShouldSerializeIsNew()
		{
			return true;
		}

		public virtual bool ShouldSerializeTaxonomy()
		{
			return true;
		}

		public virtual bool ShouldSerializeMenu()
		{
			return true;
		}

		public virtual bool ShouldSerializeMetaDataFieldValues()
		{
			return true;
		}

        public override string GetPermissionEntityId()
        {
            return string.Format("{0}-{1}", Id, LCID);
        }

        public static int Compare([AllowNull] Content x, [AllowNull] Content y)
        {
            if (x == null)
            {
                return 1;
            }

            if (y == null)
            {
                return 1;
            }

            return x._dateCreated.CompareTo(y._dateCreated);
        }

        public int CompareTo([AllowNull] Content other)
        {
            if(other == null)
            {
                return 1;
            }

            return _dateCreated.CompareTo(other._dateCreated);
        }
        #endregion
    }
}

File: MD.CMS.Administration/MD.CMS.Administration.Core.Web/scripts/businessLogic/dataAccess/entities/content.ts
/// <reference path="./base/iBaseEntity.ts" />
/// <reference path="./base/baseEntity.ts" />
/// <reference path="./user.ts" />
/// <reference path="./metaDataFieldValue.ts" />
/// <reference path="./template.ts" />
/// <reference path="./taxonomy.ts" />
/// <reference path="./contentAlias.ts" />
/// <reference path="./contentTypeDefinition.ts" />
/// <reference path="./contentTypeDefinitionFieldValue.ts" />

namespace mdBusinessLogic.dataAccess.entities {
    export class content extends base.BaseEntity implements base.IBaseEntity<content> {
        public LCID: number;
        public DateCreated: Date;
        public AuthorId: number;
        public FolderId: number;
        public Title: string;
        public Path: string;
        public Html: string;
        public Author: user;
        public ContentType: contentTypeDefinition<contentTypeDefinitionFieldValue>;
        public Taxonomy: Array<taxonomy>;
        public MetaDataFieldValues: Array<metaDataFieldValue>;
        public ContentAliases: Array<contentAlias>;
        public Template: template;
        public IsNew: boolean;
        public IsPublished: boolean;
        public IsDataBound: boolean;
        public UniqueId: string;
        public ContentTypeDefinitionId: number;

        constructor(obj?: content) {
            super(obj);
            this.LCID = 0;
            this.DateCreated = new Date();
            this.AuthorId = 0;
            this.FolderId = 0;
            this.Title = "";
            this.Path = "";
            this.Html = null;
            this.Author = null;
            this.ContentType = null;
            this.Taxonomy = new Array<taxonomy>();
            this.MetaDataFieldValues = new Array<any>();
            this.ContentAliases = new Array<contentAlias>();
            this.Template = null;
            this.IsNew = true;
            this.IsPublished = false;
            this.IsDataBound = false;
            this.UniqueId = "";
            this.ContentTypeDefinitionId = 0;

            if (obj !== undefined && obj != null) {
                this.construct(obj);
            }
        }

        public construct(data: any) {
            super.construct(data);
            this.LCID = this.getValue<number>(data, 'LCID', 0);
            this.DateCreated = this.getValue<Date>(data, 'DateCreated', null);
            this.AuthorId = this.getValue<number>(data, 'AuthorId', 0);
            this.FolderId = this.getValue<number>(data, 'FolderId', 0);
            this.Title = this.getValue<string>(data, 'Title', '');
            this.Path = this.getValue<string>(data, 'Path', '');
            this.Html = this.getValue<string>(data, 'Html', '');
            this.ContentTypeDefinitionId = this.getValue<number>(data, 'ContentTypeDefinitionId', 0);
            this.Author = this.getConstructEntityValue<user>(data, 'Author', new user());
            this.ContentType = this.getConstructEntityValue<contentTypeDefinition<contentTypeDefinitionFieldValue>>(data, 'ContentType', new contentTypeDefinition<contentTypeDefinitionFieldValue>(new contentTypeDefinitionFieldValue()));
            this.Taxonomy = this.getArrayConstructEntityValue<taxonomy>(data, 'Taxonomy', new Array<taxonomy>(), new taxonomy());
            this.MetaDataFieldValues = this.getArrayConstructEntityValue<metaDataFieldValue>(data, 'MetaDataFieldValues', new Array<metaDataFieldValue>(), new metaDataFieldValue());
            if (data.ContentAliases !== undefined && data.ContentAliases != null && data.ContentAliases.length > 0 && (typeof data.ContentAliases[0] === 'string' || data.ContentAliases[0] instanceof String)) {
                let thisObj = this;
                data.ContentAliases = data.ContentAliases.map(function (alias) {
                    let al: contentAlias = new contentAlias();
                    al.construct({
                        LCID: thisObj.LCID,
                        DateCreated: thisObj.DateCreated,
                        ContentId: thisObj.Id,
                        Alias: alias
                    });
                    return al;
                });
            }
            this.ContentAliases = this.getArrayConstructEntityValue<contentAlias>(data, 'ContentAliases', new Array<contentAlias>(), new contentAlias());
            this.Template = this.getConstructEntityValue<template>(data, 'Template', new template());
            this.IsNew = this.getValue<boolean>(data, 'IsNew', false);
            this.IsPublished = this.getValue<boolean>(data, 'IsPublished', false);
            this.IsDataBound = this.getValue<boolean>(data, 'IsDataBound', false);
            this.UniqueId = this.getValue<string>(data, 'UniqueId', '');
        }

        public clone(): content {
            return new content(this);
        }
    }
}
